color_psnr_comparison.m

```diff
@@ -0,0 +1,163 @@
+    'l', 'c', 'u', 'm', 'w', 'f', 'g', 'y', 'p', 'b', 'v', 'k', 'j', ...
+    'x', 'q', 'z', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0'};
+
+% Create decimal values for each letter, where most frequent letters get 
+% the smallest values.
+offsets = (0:length(letter_freq)-1);
+
+% Convert the decimal offsets to binary. Use base 6 since we have >32 values
+binary_offsets = num2cell(dec2bin(offsets, 6), 2)';
+
+% Map letters -> binary
+letter_map = containers.Map(letter_freq, binary_offsets);
+
+% Cast the phrase to lowercase to maximize map coverage
+wm = lower(wm);
+
+% Convert the watermark phrase to a binary sequence
+wm_bin_seq = [];
+for i = 1:length(wm)
+    if letter_map.isKey(wm(i))
+        % Append the binary value of the character to the sequence
+        wm_bin_seq = [wm_bin_seq letter_map(wm(i))];
+    else
+        % Append a ' ' if the character isn't in our map
+        wm_bin_seq = [wm_bin_seq letter_map(' ')];
+    end
+end
+
+% Convert the sequence to a GF(2) polynomial (Galois Field, you dingus)
+wm_bin_seq = str2num(reshape(wm_bin_seq, 1, [])')';
+
+% Determine BCH correction code
+% Have to pad up to valid (n,k) tuples for BCH codes
+% See: https://www.mathworks.com/help/comm/ref/bchenc.html
+wm_bin_seq = [wm_bin_seq zeros(1, bch_k - length(wm_bin_seq))];
+wm_bch = bchenc(gf(wm_bin_seq, 1), bch_n, bch_k);
+
+
+% Determine the set of coordinates that correspond to the top-left corner
+% of each DCT block. The top left corner contains the largest DCT values
+% and inserting the watermark into those spots minimizes the deterioration
+% of image quality and increase of image size by maintaining the number of
+% zero coefficients.
+% YCbCr and RGB image are same size so we can reuse the points.
+[Xs, Ys] = meshgrid((1:8:size(im_Y_dct,1)), (1:8:size(im_Y_dct,2)));
+corners = [Xs(:) Ys(:)];
+
+% Add the binary values to each block
+for i = 1:length(wm_bch)
+    if wm_bch(i) == 1
+        offset = 2;
+    else
+        offset = 0;
+    end
+
+    coord = corners(i,:);
+    x = coord(1,1); y = coord(1,2);
+    im_Y_dct(x,y)  = im_Y_dct(x,y)  + offset;
+    im_Cb_dct(x,y) = im_Cb_dct(x,y) + offset;
+    im_Cr_dct(x,y) = im_Cr_dct(x,y) + offset;
+    im_R_dct(x,y)  = im_R_dct(x,y)  + offset;
+    im_G_dct(x,y)  = im_G_dct(x,y)  + offset;
+    im_B_dct(x,y)  = im_B_dct(x,y)  + offset;
+end
+
+% Now using the watermark added DCT coefficients, construct and output the
+% watermarked image as Watermarked.tiff
+im_Y_idct  = blockproc(im_Y_dct,  [8 8], @(b) idct2(b.data));
+im_Cb_idct = blockproc(im_Cb_dct, [8 8], @(b) idct2(b.data));
+im_Cr_idct = blockproc(im_Cr_dct, [8 8], @(b) idct2(b.data));
+im_R_idct  = blockproc(im_R_dct,  [8 8], @(b) idct2(b.data));
+im_G_idct  = blockproc(im_G_dct,  [8 8], @(b) idct2(b.data));
+im_B_idct  = blockproc(im_B_dct,  [8 8], @(b) idct2(b.data));
+
+% Copy original source file to preserve all tiff metadata. It's a pain to write by hand..
```

Comment:
The Matlab folks, uh, expect you to write this all by hand

[link](https://www.mathworks.com/help/matlab/import_export/exporting-to-images.html)